### DISCUSSION TITLE
Can the fixed value BC be set in thermal-hydraulic module？

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#23140] [Q&A Modules: General] [answered]
Hi all,

I know the boundary condition in thermal-hydraulic module (THM) is inherited from the IntergralBC of MOOSE. So we need to define the flux at the boundary, which may cause the variable value not to be fixed at the boundary. Can the fixed value BC (variable value at BC is a given constant value) be used in THM?

Thank you!
Best regard

**Comment by SomnusYu** [comment]
Is there anyone kindly tell me？
Thank you.

**Reply by GiudGiud** [reply]
@joshuahansel 
I looked through the tests in modules/thermalhydraulics and didnt see any DirichletBCs on fluid variables.

**Reply by joshuahansel** [reply]
That's correct.

**Comment by joshuahansel** [comment, marked as answer]
There aren't any components that use Dirichlet BCs on any fluid variables - we universally use weak, "ghost cell" boundary conditions. I don't remember exactly why this is the case:
- The fluid variables are `CONST MONOMIAL`, so their values live at element centers, not nodes. I don't know if Dirichlet BCs are possible for elemental variables (@GiudGiud?). I believe we at least decided that it didn't make much sense to prescribe the boundary element center values.
- The other equations would probably have their BCs computed weakly, and it would need to be consistent with the Dirichlet BC value. This may be difficult/impossible.

**Reply by SomnusYu** [reply]
Thanks for your reply. 

>  I don't know if Dirichlet BCs are possible for elemental variables.

The DirichletBC is not available for elemental variables. Does that mean THM can't set fix value BC?  If we want to let the variable value of the first element as far as possible equal to what we set in the ghost cell, the only way is to refine the grid.

Thank you!

**Reply by joshuahansel** [reply]
I believe that's correct.

**Reply by SomnusYu** [reply]
Hi joshuahansel，

Thank you for your reply. Could you kindly answer another question I have? 
In ADBoundaryFlux3EqnGhostBase.C, the ghost cell variables are calculated by the type of boundary and the variables in the interior cell. And then the boundary flux is calculated by the variable in the ghost and interior cell according to the chosen numerical method. What I want to know is why the left boundary and right boundary can use the same expression.
( **flux = _numerical_flux.getFlux(iside, ielem, true, U1, U2, normal(0));**)

In the right boundary, the ghost cell is on the right and the interior cell is on the left. This expression is exactly correct. However, in the left boundary, the ghost cell is on the left and the interior cell is on the right, so I think it should be 
( **flux =_numerical_flux.getFlux(iside, ielem, false, U2, U1, normal(0));**)

It really confused me. Could you tell me the reason?

Thank you very much!!

**Reply by joshuahansel** [reply]
Good question. Basically, first the flux is computed in a reference space with the first argument on the left and the second argument on the right, and then it is transformed using `normal(0)`, which is equal to +1 for the case that the ordering is left, right, and it's -1 for right, left.

**Reply by SomnusYu** [reply]
Thank you! Where did the transformation be executed? In the numerical method that we chose (ADNumericalFlux3EqnHLLC)? And if we don't use the transformation in the numerical method, should the expression be the following?

**if (normal(0) == -1)
    flux = _numerical_flux.getFlux(iside, ielem, false, U2, U1, normal(0));
else if (normal(0) == 1)
    flux = _numerical_flux.getFlux(iside, ielem, true, U1, U2, normal(0));**

Thank you!

**Reply by joshuahansel** [reply]
That's right, it's in `ADNumericalFlux3EqnHLLC`. And yes, if you don't have it in there, what you have is correct, other than that you shouldn't check equality of floating point numbers; you should do `MooseUtils::absoluteFuzzyEqual(normal(0), 1.0)` for example.

**Reply by SomnusYu** [reply]
Thank you very much! Much appreciate!
